controllers/rosacluster_controller.go

```diff
@@ -111,13 +133,78 @@ func (r *ROSAClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		return reconcile.Result{}, fmt.Errorf("failed to patch ROSACluster: %w", err)
 	}
 
+	if controlPlane.Status.Ready {
+		rosaScope, err := scope.NewROSAControlPlaneScope(scope.ROSAControlPlaneScopeParams{
+			Client:         r.Client,
+			Cluster:        cluster,
+			ControlPlane:   controlPlane,
+			ControllerName: "",
+			Logger:         log,
+			NewStsClient:   r.NewStsClient,
+		})
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to create rosa controlplane scope: %w", err)
+		}
+
+		if r.NewOCMClient == nil {
+			return ctrl.Result{}, fmt.Errorf("failed to create OCM client: NewOCMClient is nil")
+		}
+
+		ocmClient, err := r.NewOCMClient(ctx, rosaScope)
+		if err != nil || ocmClient == nil {
+			return ctrl.Result{}, fmt.Errorf("failed to create OCM client: %w", err)
+		}
+
+		// List the ROSA-HCP nodePools and MachinePools
+		nodePools, err := ocmClient.GetNodePools(rosaScope.ControlPlane.Status.ID)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to get nodePools: %w", err)
+		}
+
+		rosaMPNames, err := r.getRosaMachinePoolNames(ctx, cluster)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to get Rosa machinePool names: %w", err)
+		}
+
+		// Ensure every NodePool has a MachinePool and create a corresponding MachinePool if it does not exist.
+		var errs []error
+		for _, nodePool := range nodePools {
+			// continue if nodePool is not in ready state.
+			if !rosa.IsNodePoolReady(nodePool) {
+				continue
+			}
+			// continue if nodePool exist
+			if rosaMPNames[nodePool.ID()] {
+				continue
+			}
+			// create ROSAMachinePool & MachinePool
+			rosaMachinePool, machinePool := r.buildROSAMachinePool(nodePool.ID(), cluster.Name, cluster.Namespace, nodePool)
+
+			log.Info(fmt.Sprintf("Create ROSAMachinePool %s", rosaMachinePool.Name))
+			if err = r.Client.Create(ctx, rosaMachinePool); err != nil {
+				errs = append(errs, err)
+			}
+
+			log.Info(fmt.Sprintf("Create MachinePool %s", machinePool.Name))
+			if err = r.Client.Create(ctx, machinePool); err != nil {
+				errs = append(errs, err)
+			}
+		}
+
+		if len(errs) > 0 {
+			return ctrl.Result{}, kerrors.NewAggregate(errs)
+		}
+	}
+
```

Comment:
I think this entire block could be extracted out to a function with a good name and that simplifies reading this and gaining context without delving into the details

Reply:
okay, done